File: solutions/lua/knapsack/1/knapsack.lua
return {
	maximum_value = function(maximum_weight, items)
		local buffer = {}
		for i = 0, maximum_weight do
			buffer[i] = 0
		end
		for _, item in ipairs(items) do
			local i = maximum_weight + 1
			while i > item.weight do
				i = i - 1
				buffer[i] = math.max(buffer[i], item.value + buffer[i - item.weight])
			end
		end
		return buffer[maximum_weight]
	end,
}

File: solutions/ruby/knapsack/1/knapsack.rb
# frozen_string_literal: true

# class Knapsack
class Knapsack
  def initialize(max_weight) = @max_weight = max_weight

  def max_value(items)
    dp = Array.new(@max_weight + 1, 0)
    items.each do |item|
      @max_weight.downto(item.weight) do |capacity|
        dp[capacity] = [dp[capacity], dp[capacity - item.weight] + item.value].max
      end
    end
    dp[@max_weight]
  end
end
